### ISSUE TITLE
spelling error in _sslverify.OpenSSLCertificateOptions.__getstate__()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2045]
|<img alt="dq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dq reported|
|-|-|
|Trac ID|trac#2045|
|Type|defect|
|Created|2006-09-01 13:27:51Z|

the line 881 in file _sslverify.py
```python
del d['context']
```
should be written
```python
del d['_context']
```

<details><summary>Searchable metadata</summary>

```
trac-id__2045 2045
type__defect defect
reporter__dq dq
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1157117271000000 1157117271000000
changetime__1157565415000000 1157565415000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

There are actually numerous problems with this area of OpenSSLCertificateOptions.  Here's a patch to test and fix them:

```
Index: twisted/test/test_sslverify.py   
===================================================================
--- twisted/test/test_sslverify.py      (revision 17970)
+++ twisted/test/test_sslverify.py      (working copy)
@@ -9,6 +9,7 @@
 from twisted.trial import unittest
 from twisted.internet import protocol, defer, reactor
 from twisted.python import log
+from twisted.python.reflect import objgrep, isSame

 from twisted.internet import _sslverify as sslverify

@@ -174,6 +175,43 @@
         self.assertIn('Locality Name', s)


+    def test_certificateOptionsSerialization(self):
+        """
+        Test that __setstate__(__getstate__()) round-trips properly.
+        """
+        firstOpts = sslverify.OpenSSLCertificateOptions(
+            privateKey=self.sKey,
+            certificate=self.sCert,
+            method=SSL.SSLv3_METHOD,
+            verify=True,
+            caCerts=[self.sCert],
+            verifyDepth=2,
+            requireCertificate=False,
+            verifyOnce=False,
+            enableSingleUseKeys=False,
+            enableSessions=False,
+            fixBrokenPeers=True)
+        context = firstOpts.getContext()
+        state = firstOpts.__getstate__()
+        
+        # The context shouldn't be in the state to serialize
+        self.failIf(objgrep(state, context, isSame), objgrep(state, context, isSame))
+
+        opts = sslverify.OpenSSLCertificateOptions()
+        opts.__setstate__(state)
+        self.assertEqual(opts.privateKey, self.sKey)
+        self.assertEqual(opts.certificate, self.sCert)
+        self.assertEqual(opts.method, SSL.SSLv3_METHOD)
+        self.assertEqual(opts.verify, True)
+        self.assertEqual(opts.caCerts, [self.sCert])
+        self.assertEqual(opts.verifyDepth, 2)
+        self.assertEqual(opts.requireCertificate, False)
+        self.assertEqual(opts.verifyOnce, False)
+        self.assertEqual(opts.enableSingleUseKeys, False)
+        self.assertEqual(opts.enableSessions, False)
+        self.assertEqual(opts.fixBrokenPeers, True)
+        
+
     def testAllowedAnonymousClientConnection(self):
         onData = defer.Deferred()
         self.loopback(sslverify.OpenSSLCertificateOptions(privateKey=self.sKey, certificate=self.sCert, requireCertificate=False),
Index: twisted/internet/_sslverify.py
===================================================================
--- twisted/internet/_sslverify.py      (revision 17970)
+++ twisted/internet/_sslverify.py      (working copy)
@@ -876,14 +876,18 @@


     def __getstate__(self):
-        d = super(OpenSSLCertificateOptions, self).__getstate__()
+        d = self.__dict__.copy()
         try:
-            del d['context']
+            del d['_context']
         except KeyError:
             pass
         return d


+    def __setstate__(self, state):
+        self.__dict__ = state
+
+
     def getContext(self):
         """Return a SSL.Context object.
         """
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_2"></a>|@jerub set owner to @exarkun|
|-|-|

Thankyou. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [18060]) Apply patch from [[#2045](https://github.com/twisted/twisted/issues/2045)](https://github.com/twisted/twisted/issues/2045) - test/fix OpenSSLCertificateOptions serialization

Author: exarkun
Reviewer: jerub
Fixes #2045

This adds tests for OpenSSLCertificateOptions' __getstate__ and __setstate__, and
fixes __getstate__ to actually remove the cached context object.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
